pkg/network/ovn_kubernetes.go

```diff
@@ -345,7 +345,7 @@ func renderOVNKubernetes(conf *operv1.NetworkSpec, bootstrapResult *bootstrap.Bo
 	}
 
 	data.Data["IP_FORWARDING_MODE"] = operv1.IPForwardingRestricted
-	if c.GatewayConfig != nil {
+	if c.GatewayConfig != nil && c.GatewayConfig.IPForwarding != "" {
```

Comment:
This doesn't allow `IP_FORWARDING_MODE` to become `""` even if `c.GatewayConfig.IPForwarding` is set to it, but the validation allows IPForwarding to be set to `""`.
Additionally when a cluster with `IPForwarding=""` upgrades to this PR the `IP_FORWARDING_MODE` will change from `""` to `IPForwardingRestricted`
Is this behavior expected?

Reply:
Correct, the idea is that "IP_FORWARDING_MODE" is either "Global", or "Restricted", because these are the only modes that the bash script should be populated with. 

The new API validation allows the following 4 input scenarios:
i) accept old invalid input
ii) accept c.GatewayConfig with IPForwading NOT set
iii) accept c.GatewayConfig.IPForwarding = "Global"
iv) accept c.GatewayConfig.IPForwarding = "Restricted"

IP_FORWARDING_MODE is the value that's set in bindata/network/ovn-kubernetes/common/008-script-lib.yaml:
```
568       # If IP Forwarding mode is global set it in the host here.
569       ip_forwarding_flag=                                                     
570       if [ "{{.IP_FORWARDING_MODE}}" == "Global" ]; then       
571         sysctl -w net.ipv4.ip_forward=1          
572         sysctl -w net.ipv6.conf.all.forwarding=1
573       else                                   
574         ip_forwarding_flag="--disable-forwarding"
575       fi 
```

Read line 347 and line 348:
```
        data.Data["IP_FORWARDING_MODE"] = operv1.IPForwardingRestricted
	if c.GatewayConfig != nil && c.GatewayConfig.IPForwarding != "" {
          data.Data["IP_FORWARDING_MODE"] = c.GatewayConfig.IPForwarding
       }
```

Scenario:
i) accept old invalid input:
    data.Data["IP_FORWARDING_MODE"] = <old invalid input>
    pod will not be redeployed, line 570 of the script will read:
         if [ "<old invalid value>" == "Global" ]; then     ----> Restricted
ii) accept c.GatewayConfig with IPForwading NOT set
    data.Data["IP_FORWARDING_MODE"] = "Restricted"
    pod will be redeployed if IP_FORWARDING_MODE had a different value from "Restricted"
    line 570 of the script will read:
         if [ "Restricted" == "Global" ]; then     ----> Restricted
iii) accept c.GatewayConfig.IPForwarding = "Global"
     data.Data["IP_FORWARDING_MODE"] = Global
     pod will be redeployed if IP_FORWARDING_MODE had a different value from "Global"
     line 570 of the script will read:
         if [ "Global" == "Global" ]; then     ----> Global
iv) accept c.GatewayConfig.IPForwarding = "Restricted"
     data.Data["IP_FORWARDING_MODE"] = Restricted
     pod will be redeployed if IP_FORWARDING_MODE had a different value from "Restricted"
     line 570 of the script will read:
         if [ "Restricted" == "Global" ]; then     ----> Restricted

The point is that there's no difference between `c.GatewayConfig with IPForwading NOT set` and between `c.GatewayConfig.IPForwarding = "Restricted"`, both should yield `data.Data["IP_FORWARDING_MODE"] = Restricted` and both should give `if [ "Restricted" == "Global" ]; then` in the bash script. Otherwise, the pods will be redeployed without any reason if the admin goes from `c.GatewayConfig.IPForwarding = "Restricted"` to  `c.GatewayConfig with IPForwading NOT set` or vice versa

Reply:
> Additionally when a cluster with IPForwarding="" upgrades to this PR the IP_FORWARDING_MODE will change from "" to IPForwardingRestricted

Yeah, that's the unfortunate side effect; but it should only happen once on upgrade, and then never again. The alternative is to have it in the reverse way and to have it set to `""`. Which might be better as the default is Restricted not set in the script?

Becaues you're right, the current default configuration is:
```
      # If IP Forwarding mode is global set it in the host here.
      ip_forwarding_flag=
      if [ "" == "Global" ]; then
        sysctl -w net.ipv4.ip_forward=1
        sysctl -w net.ipv6.conf.all.forwarding=1
      else
        ip_forwarding_flag="--disable-forwarding"
      fi
```

Reply:
@kyrtapz I switched the logic around now:
```
// We accept 3 valid inputs:
	// c.GatewayConfig.IPForwarding not set --> defaults to "".
	// c.GatewayConfig.IPForwarding set to "Restricted".
	// c.GatewayConfig.IPForwarding set to "Global".
	// For "" and "Restricted" (which behave the exact same), the ConfigMap's .IP_FORWARDING_MODE shall be "".
	// For "Global", the ConfigMap's .IP_FORWARDING_MODE shall be "Global".
	data.Data["IP_FORWARDING_MODE"] = ""
	if c.GatewayConfig != nil && c.GatewayConfig.IPForwarding == operv1.IPForwardingGlobal {
		data.Data["IP_FORWARDING_MODE"] = c.GatewayConfig.IPForwarding
	}
```

That will:
* redeploy all pods during an upgrade if an invalid value was used, and it will populate the CM with ""
* redeploy all pods if "Restricted" was explicitly set  in the CM (unfortunately, that's just what it is .. it's either, or) so these pods will redeploy once on upgrade .... but then never again, because the CM will never again be set with "Restricted"
* keep all pods unchanged if the CM contains "" (which should be the vast majority of deployments, as this was the default value)

The idea is that if I as an admin switch between  c.GatewayConfig.IPForwarding " Restricted" and between  c.GatewayConfig.IPForwarding not being present, the pods shouldn't redeploy, as both should set the CM to ""; pods also shouldn't redeploy if I manage to sneak through API validation and insert an invalid value, it the CM should have "". Pods should only redeploy if I change from or to "Global". Does that make sense?

Reply:
Thanks!
The configmap is applied before the ovn-k manifests so it shouldn't affect upgrades anyway.

> keep all pods unchanged if the CM contains "" (which should be the vast majority of deployments, as this was the default value)

this is only the case when the `GatewayConfig` is not nil, I don't think it is the majority.